compiler/build.gradle

```diff
@@ -8,15 +9,117 @@ description = 'The protoc plugin for gRPC Java'
 buildscript {
     repositories {
         mavenCentral()
+        mavenLocal()
     }
     dependencies {
-        classpath libraries.protobuf_plugin
+        classpath libraries.protobuf_plugin,
+                  'com.google.gradle:osdetector-gradle-plugin:1.2.1'
     }
 }
 
-executables {
-  java_plugin {
-    baseName "$protocPluginBaseName"
+// We know how to build deployment artifacts from these host architectures
+def deployableHostArchs = ['x86_32', 'x86_64'] as HashSet
+
+// When there is only one platform available, Gradle doesn't create a directory
+// for the sole platform. In order to keep the script simple, we intentionally
+// always build the 'local_arch' even though it's duplicate with one of the
+// deployableHostArchs, so that we always have at least two platforms.
+def targetArchs = ['local_arch'] as HashSet
+
+def artifactStagingPath = "$buildDir/artifacts" as File
+def artifactPath = { arch ->
+  return "$artifactStagingPath/java_pluginExecutable/" + arch + "/${protocPluginBaseName}.exe"
+}
+
+
+if (System.env.TARGET_ARCHS != null) {
+  def archs = System.env.TARGET_ARCHS.split(' +')
+  targetArchs.addAll(archs)
+} else {
+  targetArchs.add(osdetector.arch)
+}
+
+// Adds space-delimited arguments from the environment variable env to the
+// argList.
+def addEnvArgs = { env, argList ->
+  def value = System.getenv(env)
+  if (value != null) {
+    value.split(' +').each() { it -> argList.add(it) }
+  }
+}
+
+model {
+  toolChains {
+    gcc(Gcc) {
+      target("x86_64") {
+        cppCompiler.withArguments { args ->
+          args << "-m64"
+        }
+        linker.withArguments { args ->
+          args << "-m64"
+        }
+      }
+      target("x86_32") {
+        cppCompiler.withArguments { args ->
+          args << "-m32"
+        }
+        linker.withArguments { args ->
+          args << "-m32"
+        }
+      }
+      target('local_arch') { }
+    }
+    clang(Clang) {
+      target("x86_64") {
+        cppCompiler.withArguments { args ->
+          args << "-m64"
+        }
+        linker.withArguments { args ->
+          args << "-m64"
+          // Usually we need to pass "-L" args to set library search path.
+          // Clang ld has an issue when multiple versions (e.g., 32-bit and
+          // 64-bit) of the same library exist in the search paths. It will
+          // only look at the first one even if the correct one comes after.
+          // So we have to specify paths for different architectures
+          // separately through separate flags.
+          addEnvArgs("LDFLAGS_x86_64", args)
```

Comment:
We only care about addEnvArgs for OS X. Do we even care about 32bit on that platform? Snow Leopard (10.6) was the last version of OS X to support 32 bit, and it is unsupported as of February 2014.